zipline/data/adjustments.py

```diff
@@ -349,72 +350,62 @@ def calc_dividend_ratios(self, dividends):
-
-        effective_dates = np.full(len(amounts), -1, dtype=int64_dtype)
-
-        calendar = self._calendar
-
-        # Calculate locs against a tz-naive cal, as the ex_dates are tz-
-        # naive.
-        #
-        # TODO: A better approach here would be to localize ex_date to
-        # the tz of the calendar, but currently get_indexer does not
-        # preserve tz of the target when method='bfill', which throws
-        # off the comparison.
-        tz_naive_calendar = calendar.tz_localize(None)
-        day_locs = tz_naive_calendar.get_indexer(ex_dates, method='bfill')
-
-        isnull = pd.isnull
-
-        for i, amount in enumerate(amounts):
-            sid = sids[i]
-            ex_date = ex_dates[i]
-            day_loc = day_locs[i]
-
-            prev_close_date = calendar[day_loc - 1]
-
-            try:
-                prev_close = equity_daily_bar_reader.get_value(
-                    sid, prev_close_date, 'close')
-                if not isnull(prev_close):
-                    ratio = 1.0 - amount / prev_close
-                    ratios[i] = ratio
-                    # only assign effective_date when data is found
-                    effective_dates[i] = ex_date
-            except NoDataOnDate:
-                log.warn("Couldn't compute ratio for dividend %s" % {
-                    'sid': sid,
-                    'ex_date': ex_date,
-                    'amount': amount,
-                })
-                continue
-
-        # Create a mask to filter out indices in the effective_date, sid, and
-        # ratio vectors for which a ratio was not calculable.
-        effective_mask = effective_dates != -1
-        effective_dates = effective_dates[effective_mask]
-        effective_dates = effective_dates.astype('datetime64[ns]').\
-            astype('datetime64[s]').astype(uint32_dtype)
-        sids = sids[effective_mask]
-        ratios = ratios[effective_mask]
 
+        pricing_reader = self._equity_daily_bar_reader
+        input_sids = dividends.sid.values
+        unique_sids, sids_ix = np.unique(input_sids, return_inverse=True)
+        dates = pricing_reader.sessions.values
+
+        close, = pricing_reader.load_raw_arrays(
+            ['close'],
+            pd.Timestamp(dates[0], tz='UTC'),
+            pd.Timestamp(dates[-1], tz='UTC'),
+            unique_sids,
+        )
+        date_ix = np.searchsorted(dates, dividends.ex_date.values)
+        mask = date_ix > 0
+
+        date_ix = date_ix[mask]
+        sids_ix = sids_ix[mask]
+        input_dates = dividends.ex_date.values[mask]
+
+        # subtract one day to get the close on the day prior to the merger
+        previous_close = close[date_ix - 1, sids_ix]
+        input_sids = input_sids[mask]
+
+        amount = dividends.amount.values[mask]
+        ratio = 1.0 - amount / previous_close
+
+        non_nan_ratio_mask = ~np.isnan(ratio)
+        for ix in np.flatnonzero(~non_nan_ratio_mask):
```

Comment:
Do we have test coverage for these warning cases? I think we could set that up relatively easily using an in-memory bar reader.